src/audio/module_adapter/module_adapter.c

```diff
@@ -505,12 +505,12 @@ int module_adapter_copy(struct comp_dev *dev)
 	ret = module_process(mod, mod->input_buffers, mod->num_input_buffers,
 			     mod->output_buffers, mod->num_output_buffers);
 	if (ret) {
-		if (ret == -ENOSPC || ret == -ENODATA) {
-			ret = 0;
-			goto consume;
+		if (ret != -ENOSPC && ret != -ENODATA) {
+			comp_err(dev, "module_adapter_copy() error %x: module processing failed",
+				 ret);
+			goto out;
 		}
-		comp_err(dev, "module_adapter_copy() error %x: module processing failed", ret);
-		goto out;
+		ret = 0;
```

Comment:
doesn't this not only simplify but also change the logic? Before this patch in one of `-ENOSPC` or `-ENODAT` cases the below copying loop would be skipped, now it is executed. Is this allowed because the loop will do nothing in such a case? If so at least a comment would be good. Otherwise, if the original logic was correct, maybe just put the below loop in an `else` clause?

Reply:
@ranj063 could you explain a bit more, please - that means that `.size` will be equal to 0 below, right? If so, I doubt that this is an optimisation. Your PR is aiming at performance optimisations. Here you replace a single jump with potentially multiple ones - the `list_for_item()` loop, plus a pointer calculation, unless the compiler moves it under the `if` below. And a couple more calculations to obtain `.size`. I agree that it's better not to have `goto`s in the middle of the code, but this seems a rather high price to pay on a hot path. How about
```
	if (ret == -ENOSPC || ret == -ENODATA) {
		ret = 0;
	} else if (ret) {
		comp_err(dev, "module_adapter_copy() error %x: module processing failed",
			 ret);
		goto out;
	} else {
		/* copy all produced output samples to output buffers */
		i = 0;
		list_for_item(blist, &mod->sink_buffer_list) {
			struct comp_buffer *buffer = container_of(blist, struct comp_buffer, sink_list);

			if (mod->output_buffers[i].size > 0) {
				ca_copy_from_module_to_sink(&buffer->stream, mod->output_buffers[i].data,
							    mod->output_buffers[i].size);
				audio_stream_produce(&buffer->stream, mod->output_buffers[i].size);
			}
			i++;
		}
	}
```

Reply:
@lyakh the main reason for moving this is really the optimization in the 4th patch in the series. I understand this patch itself offers no performance optimization. What I really want to do is avoid the intermediate copy before consuming as is done today